scheduler: bugfix: Make pod_scenario_builder build scenarios for rest of elastic job

ArmedGuy opened the pull request:
While going through the solvers, @bugra-gedik-nv noticed a potential bug in pod_scenario_builder, where we jump to the next scenario if any of the potentialVictimTasks are already in recordedVictimTasks.
https://github.com/NVIDIA/KAI-Scheduler/blob/main/pkg/scheduler/actions/common/solvers/pod_scenario_builder.go#L77-L82

When an elastic job is the victim, `GetTasksToEvict` will only return the last task from reverse TaskOrderFn of a victim until the job has equal or less than `MinAvailable` tasks to consider for scenario (the `jobHasMoreTasks = false` situation).

https://github.com/NVIDIA/KAI-Scheduler/blob/main/pkg/scheduler/actions/common/solvers/pod_scenario_builder.go#L71-L82

When you have a reclaimer/preemptor with at least 2 tasks to schedule and no free nodes, we would successfully solve for the first task in job_solver.go by calling `solvePartialJob`, and `recordedVictimJobs` is now populated with an elastic job representation with 1 task. But when solving for 2 tasks in the next iteration of `solvePartialJob`,  the first call to `GetNextScenario()` with `recordedVictimJobs` populated  with the elastic job representation with 1 pod would match the first pod in `potentialVictimTasks` from `GetTasksToEvict()` and directly call  `GetNextScenario()` internally. 
However, because we already popped the full elastic job off `victimsJobsQueue`, the next scenario will no longer evaluate any of the other pods in that elastic job and move on to another job in the victim queue.

In the scenario where the elastic job is the only victim, no further scenario is presented and solver would consider the scenario unsolvable even though the elastic job victim could give up more tasks.

More generalized: it would mean elastic jobs can currently only have 1 pod preempted each scheduling loop.

The proposed solution is when a victimJob popped of the queue has any task that is already part of reportedVictims, we construct a job representative with all remaining tasks that are not reportedVictims, and if the number of remainingTasks is non-zero we push that job back onto the victimQueue before calling GetNextScenario.

ArmedGuy commented:
Oh and also re-added ginkgo test suite needed to run these tests, as coverage was missing.

bugra-gedik-nv commented on pkg/scheduler/actions/common/solvers/pod_scenario_builder_test.go:
```diff
@@ -142,6 +143,53 @@ var _ = Describe("PodAccumulatedScenarioBuilder", func() {
+			// Only the first pod group with the last task is recordedVictimJobs
+			for _, podGroupInfo := range ssn.PodGroupInfos {
+				if podGroupIndex == 0 {
+					podIndex := 0
```
podIndex unused?

bugra-gedik-nv commented on pkg/scheduler/actions/common/solvers/pod_scenario_builder_test.go:
```diff
@@ -142,6 +143,53 @@ var _ = Describe("PodAccumulatedScenarioBuilder", func() {
+			// Only the first pod group with the last task is recordedVictimJobs
+			for _, podGroupInfo := range ssn.PodGroupInfos {
+				if podGroupIndex == 0 {
+					podIndex := 0
```
podIndex unused?

ArmedGuy replied:
good flag, refactored it out before as maps arent ordered. will remove

enoodle replied:
I wonder why the linter didn't fail on it

bugra-gedik-nv replied:
Strictly speaking, the variable is used down below (it is incremented). However it is not needed. Perhaps the use makes it pass linting.

ArmedGuy replied:
yep the incr caused linter to not care, I removed this.

bugra-gedik-nv commented on pkg/scheduler/actions/common/solvers/pod_scenario_builder_test.go:
```diff
@@ -142,6 +143,53 @@ var _ = Describe("PodAccumulatedScenarioBuilder", func() {
+			podGroupIndex := 0
+			// Only the first pod group with the last task is recordedVictimJobs
+			for _, podGroupInfo := range ssn.PodGroupInfos {
+				if podGroupIndex == 0 {
```
nit: do we need a for loop here? Just take the first of the pod group infos perhaps

ArmedGuy replied:
You will be surprised at how annoying it is to get the "first" value of a map. It seems it might be possible to do `vals := slices.Collect(maps.Values(m))` if the project uses a new enough Golang version (Aug 2024).

But can remove the if and just break after first element in the loop in this case.

bugra-gedik-nv commented on pkg/scheduler/actions/common/solvers/pod_scenario_builder_test.go:
```diff
@@ -142,6 +143,53 @@ var _ = Describe("PodAccumulatedScenarioBuilder", func() {
+				if podGroupIndex == 0 {
+					podIndex := 0
+					var partialTasks []*pod_info.PodInfo
+					for _, podInfo := range podGroupInfo.PodInfos {
```
nit: can we just take the last one without a for loop?

bugra-gedik-nv commented on pkg/scheduler/actions/common/solvers/pod_scenario_builder_test.go:
```diff
@@ -142,6 +143,53 @@ var _ = Describe("PodAccumulatedScenarioBuilder", func() {
+				numberOfGeneratedScenarios += 1
+			}
+
+			Expect(numberOfGeneratedScenarios).To(Equal(3))
```
Can we assert individual scenarios to make sure they are as expected?

davidLif reviewed (changes requested):
I  reproduced the issue as an e2e test - https://github.com/NVIDIA/KAI-Scheduler/tree/DavidL/reclaim-elastic-to-distributed-bug.

However, I am not sure about the suggested solution. 
The `remainingTasks` will contain all the all the elastic job pods except for the "last" pod (which is part of the recorded victims). This means that all the pods for this elastic jobs will become potential victims. In some cases (for example, if the reclaimee job has min +2 pods) we want only one pod to join as the the next "potential victim".

I will look into it a bit more, but of the top of my head an implementation of `podgroup_info.GetTasksToEvict` that knows to skip spesific pods might be more closly aligned to the required result.

Thank you for your input — this is an important bug.

bugra-gedik-nv commented:
> I reproduced the issue as an e2e test - https://github.com/NVIDIA/KAI-Scheduler/tree/DavidL/reclaim-elastic-to-distributed-bug.
> 
> However, I am not sure about the suggested solution. The `remainingTasks` will contain all the all the elastic job pods except for the "last" pod (which is part of the recorded victims). This means that all the pods for this elastic jobs will become potential victims. In some cases (for example, if the reclaimee job has min +2 pods) we want only one pod to join as the the next "potential victim".
> 
> I will look into it a bit more, but of the top of my head an implementation of `podgroup_info.GetTasksToEvict` that knows to skip spesific pods might be more closly aligned to the required result.
> 
> Thank you for your input — this is an important bug.

@davidLif For elastic jobs, GetTasksToEvict seems to return 1 pod if num pods > MinAvailable and MinAvailable pods if num pods == MinAvailable.

> This means that all the pods for this elastic jobs will become potential victims

I don't see this, but might be wrong. Can you explain how this can happen. We should be able to see it unit test.

@ArmedGuy  Would be good to have the unit test updated to MinAvailable = 2 and total pods to 4, so the possible scenarios after one pod is already in recorded victims are:
- evict 1
- evict 3

ArmedGuy commented:
@davidLif It should not be a problem, because this logic of taking all tasks that aren't reportedVictims already and pushing a job representative back into the queue with them will make the job pass by `GetTasksToEvict` on the next scenario, which does the filtering down to 1 job for us. It is the same logic that is already prescribed further down when `GetTasksToEvict` returns an elastic job with `jobHasMoreTasks = true`.

davidLif commented:
> @davidLif It should not be a problem, because this logic of taking all tasks that aren't reportedVictims already and pushing a job representative back into the queue with them will make the job pass by `GetTasksToEvict` on the next scenario, which does the filtering down to 1 job for us. It is the same logic that is already prescribed further down when `GetTasksToEvict` returns an elastic job with `jobHasMoreTasks = true`.

I stand corrected — you were right, and my earlier statement was inaccurate.

ArmedGuy commented:
Added 2 new test cases to validate potential victims returned in each scenario to address @bugra-gedik-nv s point. Added for both elastic and non-elastic to validate both.

ArmedGuy commented:
@davidLif Should I pull in the e2e tests to this diff so we can see fix passing them?

bugra-gedik-nv commented on pkg/scheduler/actions/common/solvers/pod_scenario_builder_test.go:
```diff
@@ -141,6 +142,125 @@ var _ = Describe("PodAccumulatedScenarioBuilder", func() {
+			numberOfGeneratedScenarios := 0
+			potentialVictimsPerScenario := []int{0, 2}
+			for sn := scenarioBuilder.GetCurrentScenario(); sn != nil; sn = scenarioBuilder.GetNextScenario() {
+				Expect(numberOfGeneratedScenarios < len(potentialVictimsPerScenario)).To(BeTrue())
```
better to have a final excpet outside of the loop, in case we generate less scenarios than expected.

ArmedGuy replied:
This was to prevent slice access from erroring out, giving a cryptic runtime panic.

ArmedGuy replied:
But good point on still ensuring all scenarios were validated, will keep the expect outside as well.

bugra-gedik-nv commented on pkg/scheduler/actions/common/solvers/pod_scenario_builder_test.go:
```diff
@@ -141,6 +142,125 @@ var _ = Describe("PodAccumulatedScenarioBuilder", func() {
+				Expect(len(sn.RecordedVictimsJobs())).To(Equal(len(recordedVictimsJobs)))
+				numberOfGeneratedScenarios += 1
+			}
+
```
Any reason we don't have `potentialVictimsPerScenario` for this one?

ArmedGuy replied:
I separated the unit tests so the original makes sure RecordedVictimJobs is preserved (the intent of the original unit test) and the new ones check for potentialVictims instead.

bugra-gedik-nv commented on pkg/scheduler/actions/common/solvers/pod_scenario_builder_test.go:
```diff
@@ -141,6 +142,126 @@ var _ = Describe("PodAccumulatedScenarioBuilder", func() {
 			Expect(numberOfGeneratedScenarios).To(Equal(2))
 		})
+
+		It("Returned scenarios have correct number of potential victims", func() {
```
nit: I think the test case descriptions are made to read like a sentence:

It("returns scenarios with the correct number of potential victims")

davidLif commented:
> @davidLif Should I pull in the e2e tests to this diff so we can see fix passing them?

@ArmedGuy yes.

bugra-gedik-nv commented on test/e2e/suites/reclaim/reclaim_elastic_test.go:
```diff
@@ -145,4 +145,86 @@ var _ = Describe("Reclaim with Elastic Jobs", Ordered, func() {
+		wait.ForPodsScheduled(ctx, testCtx.ControllerClient, reclaimerNamespace, reclaimerPods)
+
+		// verify results
+		wait.ForPodsWithCondition(ctx, testCtx.ControllerClient, func(watch.Event) bool {
```
don't we need to verify that the elastic reclaimee job has 1 pod remaining?

bugra-gedik-nv commented on test/e2e/suites/reclaim/reclaim_elastic_test.go:
```diff
@@ -145,4 +145,86 @@ var _ = Describe("Reclaim with Elastic Jobs", Ordered, func() {
+				LabelSelector: fmt.Sprintf("%s=%s", podGroupLabelName, reclaimeePodGroup.Name),
+			})
+			Expect(err).To(Succeed())
+			return len(pods.Items) == 0
```
should be == 1 

but then, if we want to make sure it is stable, it would be best to wait a little more to make sure it doesn't change. Not sure what's the recommended way of doing that.

ArmedGuy replied:
have an update for stability testing, but will hold that off to another PR since its a change on the entire e2e framework.
